gedi/process_l2a.py

```diff
@@ -0,0 +1,290 @@
+
+    Returns
+    -------
+    pd.DataFrame containing the user-specified contents of the h5 files for the
+    valid shots within the specified bounding box.
+    """
+    df = pd.DataFrame()
+    for _f_name in file_paths:
+        try:
+            _f = h5py.File(f"{file_dir}{_f_name}", "r")
+            print(f"Processing file {_f_name}")
+            [ul_lat, ul_lon, lr_lat, lr_lon] = bbox
+            for beam in [
+                "BEAM0000",
+                "BEAM0001",
+                "BEAM0010",
+                "BEAM0011",
+                "BEAM0101",
+                "BEAM0110",
+                "BEAM1000",
+                "BEAM1011",
+            ]:
+                tmp_df = pd.DataFrame()
+                x = _f[beam][latlayer][()]
+                y = _f[beam][lonlayer][()]
+                qual = _f[beam]["quality_flag"][()]
+
+                mask = (
+                    (qual == 1)
+                    & (x <= ul_lat)
+                    & (x >= lr_lat)
+                    & (y <= lr_lon)
+                    & (y >= ul_lon)
+                )
+                tmp_df["beam"] = [beam] * np.count_nonzero(mask)
+
+                for layer in layers:
+                    tmp_df[layer] = _f[beam][layer][()][mask].tolist()
+
+                df = df.append(tmp_df)
+
+        except KeyError as e:
+            print(f"Encountered file read error with file {_f_name}")
+
+    return df.reset_index(drop=True)
+
+
+def _compute_nan_percentile(a: np.ndarray, q: float) -> np.array:
+    """This code is taken directly from StackOverflow at:
+    https://stackoverflow.com/questions/60015245/numpy-nanpercentile-is-extremely-slow
+
+    For 2d masked array of roughly 10^5 x 100, this performs ~3x faster than
+    np.nanpercentile (~10s vs. 3s) and produces results which are identical.
+
+    Would be good to add two tests to verify: (1) the results of this function are
+    the same as np.nanpercentile, and (2) that this function takes less time on a
+    reasonably sized array, in case of future performance gains in nanpercentile.
+
+    Parameters
+    ----------
+    a : array_like
+        2d array for which to compute a row-wise percentile.
+    q : float
+        Percentile to compute -- allowed values are [0, 100]
+
+    Returns
+    -------
+    np.array of the q'th percentile for each row of a. Dimensions of [len(a), 1].
+    """
+    if q < 0 or q > 100:
+        raise ValueError(f"Expected a value between 0 and 100; received {q} instead.")
+    mask = (a >= np.nanmin(a)).astype(int)
+
+    count = mask.sum(axis=1)
+    groups = np.unique(count)
+    groups = groups[groups > 0]
+
+    p = np.zeros((a.shape[0]))
+    for g in range(len(groups)):
+        pos = np.where(count == groups[g])
+        values = a[pos]
+        values = np.nan_to_num(values, nan=(np.nanmin(a) - 1))
+        values = np.sort(values, axis=1)
+        values = values[:, -groups[g] :]
+        p[pos] = np.percentile(values, q, axis=1)
+    return p
```

Comment:
I shamelessly lifted this from the StackOverflow response cited in the docstring and took little time to sit with and understand it initially. It would not surprise me if someone who was more familiar with numpy than I was able to suggest some changes (either to improve performance or readability, for example the data are already sorted, so L137 might be able to go) here.